ADR/adr-0008-secret-mgmt-for-user-workloads.md

```diff
@@ -0,0 +1,101 @@
+
+## Decision
+
+### Architecture Overview
+#### Terminology
+
+- `UploadSecret`: A short-lived Kubernetes secret used to deliver confidential data to permanent storage and link it to the RemoteSecret CR.
+- `SecretData`: An object stored in permanent SecretStorage. Valid SecretData is always linked to a RemoteSecret CR.
+- `RemoteSecret`: A CRD that appears during upload and links `SecretData` + `DeploymentTarget(s)` + K8s `Secret`. `RemoteSecret` is linked to one (or zero) SecretData and manages its deleting/updating.
+- K8s `Secret`: What appears at the output and is used by consumers.
+- `SecretId`: A unique identifier of SecretData in permanent SecretStorage.
+- `SecretStorage`: A database eligible for storing `SecretData` (such as HashiCorp Vault, AWS Secret Manager).
+
+The idea is to introduce a new CR called `RemoteSecret`, which is a Kubernetes representation of the `K8s Secret` that is stored in permanent storage aka SecretStorage. This CR contains a reference to the destination, either a Kubernetes namespace or an `Environment`. `UploadSecret` is used to perform an upload to the permanent storage and is represented as a regular Kubernetes secret with special labels and annotations recognized by the _SPI controller_. Different implementations of `SecretStorage` can be used, such as AWS Secret Manager or HashiCorp Vault.
+
+#### Example: If the destination is `Environment`
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: jdbc-connection-parameter
+  labels:
+    spi.appstudio.redhat.com/upload-secret: secret
+    spi.appstudio.redhat.com/target-environment: prod
+type: Opaque
+data:
+ ...
+```
+```yaml
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
+metadata:
+  name: jdbc-connection-parameter
+spec:
+  secret:
+    environment: prod
+    name: jdbc-connection-parameter
+ ...
+```
+
+#### Example: If the destination is `Namespace` and secret has to be linked to SA
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: jdbc-connection-parameter
+  annotations:
+      spi.appstudio.redhat.com/secret-linkedto-serviceAccount-name: sa-star
+  labels:
+    spi.appstudio.redhat.com/upload-secret: secret
+    spi.appstudio.redhat.com/target-namespace: ns-9459039345
+type: Opaque
+data:
+ ...
+```
+```yaml
+apiVersion: appstudio.redhat.com/v1beta1
+kind: RemoteSecret
+metadata:
+  name: jdbc-connection-parameter
+spec:
+  secret:
+    namespace: ns-9459039345
+    name: jdbc-connection-parameter
+    linkedTo:
+    - serviceAccount:
+        reference:
+          name: sa-star
+```
+
+
+### Security and Access Control
+
+* Access to the secret management backend should be restricted to authorized users and components.
+* The default backend should use authentication and authorization mechanisms provided by the underlying service, such as Kubernetes or AWS IAM.
+* Workspace maintainers should have the ability to create and update secrets for their workspace on the default secret management backend.
```

Comment:
Are these the maintainers of the individual workspaces within AppStudio? Would any user within AppStudio need to have accounts created for them on the `SecretStorage` provider or would the maintainers just create the resources with k8s and the system would populate them in the backend?

Based on the consequence below, I assume "no" to the latter that the users would just create the secrets in the UI and then the system would do all of the pushing.

Reply:
Thanks. Can we just note that here as well so that we don't have to look down to the consequences to understand?